commerce/bookstore/python/load_data.py

```diff
@@ -0,0 +1,286 @@
+                                           '$p = "Technical Documentation"; $p isa genre;'
+                                           'insert $th (sub-tag: $b, sup-tag: $p) isa tag-hierarchy;')
+                transaction.commit()
+    print('Loaded genre tags.')
+    print('\nAll data loaded successful!')
+    return
+
+
+def load_data():  # Main data load function
+    with TypeDB.core_client("localhost:1729") as client:
+        with client.session(db, SessionType.DATA) as session:
+            for input in Inputs:
+                input["file"] = data_path + input["file"]
+                print("Loading from [" + input["file"] + ".csv] into TypeDB ...")
+                load_data_into_typedb(input, session)  # Main data loading function. Repeat for only file in Inputs
+            generate_ordered_items()  # Add randomly generated lists of items into orders
+            load_genre_tags()  # Load genre tags
+    return
+
+
+def check_data():  # Checking whether the DB has schema and data already
+    with TypeDB.core_client("localhost:1729") as client:
+        with client.session(db, SessionType.SCHEMA) as session:
+            with session.transaction(TransactionType.READ) as transaction:
+                try:
+                    TypeQL_read_query = 'match $b isa Book, has ISBN $x; get $x; limit 3;'
+                    transaction.query().match(TypeQL_read_query)
+                    print('The DB contains the schema and loaded data already.')
+                    return 1
+                except:  # If the attempt was unsuccessful — we consider DB as empty (brand new, no schema)
+                    return 0
+
+
+def setup():  # Loading schema
+    with TypeDB.core_client("localhost:1729") as client:
+        with client.session(db, SessionType.SCHEMA) as session:
+            with open('../schema.tql', 'r') as schema:
+                define_query = schema.read()
+                with session.transaction(TransactionType.WRITE) as transaction:
+                    try:
+                        transaction.query().define(define_query)
+                        transaction.commit()
+                        print("Loaded the " + db + " schema.")
+                        return 0
+                    except Exception as e:
+                        print('Failed to load schema: ' + str(e))
+                        return 1
+
+
+# This is a list of imported files with datasets for the DB
+Inputs = [
+    {
+        "file": "books",
+        "template": books_template
+    },
+    {
+        "file": "users",
+        "template": users_template
+    },
+    {
+        "file": "ratings",
+        "template": ratings_template
+    },
+    {
+        "file": "orders",
+        "template": orders_template
+    },
+    {
+        "file": "genres",
+        "template": genre_template
+    }
+]
+
+# This is the main body of this script
+with TypeDB.core_client("localhost:1729") as client:
```

Comment:
We have a large number of code comments here. In general, we prefer to minimise code comments, because if the code gets refactored / changed (which we do frequently), then the comments go out of date very rapidly, and people forget about them, because they're just comments - a minor thing on the side when compared against the actual source code itself.

Some of the comments are fairly clear. For example, `check the db existence` on L264 is obvious from simply reading the code, so we can just delete that comment.

On the other hand, some of the comments are necessary in order to actually make it clear what the line does. We always strive to have our code self-documenting: the code itself should describe what it does, concisely, at a high level.

For example, `if check_data() == 0`, is not clear about what it means. What are we checking about the data? What does the number "0" mean?

By reading deep into the implementation of `check_data`, we can figure out all these questions: but the average developer reading the code would be better placed to understand it if the method itself was clear in its naming and in its return value. For example, `if has_existing_data() == False` (or more tersely, `if not has_existing_data()`).

We should try to make this code more readable in the sense of "gaining a conceptual understanding, just by reading this method, of how the whole program works." Moreover, we should try to apply this throughout all of our Python code. This may be a fairly major refactor task.

Reply:
In regards to how many comments: 
I feel like we can't give too many comments here. This is not the usual code, but an example. Instead of writing some additional documentation with links to lines of code, we are using comments to annotate almost any important line. Even if it's obvious. That's important because of many reasons, including:
1. We will have a lot of different users onboarding. Including some of the juniors that might be unfamiliar with even obvious code patterns and functions. Some of them might struggle with English and additional comments in simplified English might help them.
2. This tutorial's goal is not only to lead newcomers through the process to create a working bookstore. But also to enable them to explore the solution. So the less time and effort they spend to understand the existing solution, the more time and motivation they will have to continue exploring their own ideas and variants. They can spend these additional resources to modify the tutorial a little bit for their intended use case. Or build something of their own. Even the slightest bump on the road can reduce their opinion of our product, motivation to explore it, and/or their time/resources budget for their own exploration. So with that comments even in the most obvious places, I'm trying to do some shortcuts here and there. it's not that these comments will obstruct anything. Please tell me if you think this is the case.
3. The example itself will. be mostly static. I'm thinking of adding a few addons (like debug logging), maybe even adding an automated test (to test this example with new releases of TypeDB to get a warning even if something will get broken). But in terms of refactoring, I don't think the example will change a lot over time. Even if it will change - comments are very important here and should be double-checked just like the rest of the code. 
That's why, as long as these comments are not obstructing anything, I think we should stick to this comment-rich approach in examples.

Reply:
In regards to function names - yeah, I'll try to re-read and improve what I can. I feel like there is always room for improvement. So it might be an iterative process.

Reply:
Rereading:

1. Create a Client, check if DB exists, if not, create it and load data; if it exists, connect to it.
2. If DB exists, check if it has existing data. If it does, prompt user to reload it; it not, then "setup".

The narrative is better than before - I think everything is now clear except "setup". What does setup do?

Reply:
Loads schema. I renamed the function in https://github.com/vaticle/typedb-examples/pull/143/commits/06554953e94afbf6c447c02db3032f8d63a18114